logger/handler/ringbuffer.go

```diff
@@ -0,0 +1,149 @@
+package handler
+
+import (
+	"container/ring"
+	"fmt"
+	"log"
+	"net/http"
+	"regexp"
+	"strconv"
+	"sync"
+
+	"github.com/deis/deis/logger/drain"
+	"github.com/deis/deis/logger/syslog"
+)
+
+var logStorage = make(map[string]*ring.Ring)
+var ringBufferSize int
+
+var regexpForPost = regexp.MustCompile(`^/([-a-z0-9]+)/.*`)
+var regexpForGet = regexp.MustCompile(`^/([-a-z0-9]+)/([0-9]+)/.*`)
+
+var mu sync.Mutex
+
+// Add log message to main map with ring byffers by project name
+func addToStorage(name string, message string) {
+	mu.Lock()
+	defer mu.Unlock()
+	currentRing, ok := logStorage[name]
+	if !ok {
+		r := ring.New(ringBufferSize)
+		r.Value = message
+		logStorage[name] = r
+	} else {
+		r := currentRing.Next()
+		r.Value = message
+		logStorage[name] = r
+	}
+}
+
+// Get specific amount of log lines for application name from main map with ring buffers
+func getFromStorage(name string, lines int) string {
+	mu.Lock()
+	defer mu.Unlock()
+	currentRing, ok := logStorage[name]
+	if !ok {
+		return fmt.Sprintf("Could not find logs for project '%s'", name)
+	}
+	var data string
+	getLine := func(line interface{}) {
+		if line == nil || lines <= 0 {
+			return
+		}
+		lines -= 1
+		if lines == 0 {
+			data += fmt.Sprint(line)
+		} else {
+			data += fmt.Sprintln(line)
+		}
+	}
+
+	currentRing.Next().Do(getLine)
+	return data
+}
+
+func RingBufferHandler(bufferSize int, webServicePort int) *Handler {
+	ringBufferSize = bufferSize
+	h := Handler{
+		BaseHandler: syslog.NewBaseHandler(5, filter, false),
+	}
+	go h.ringBufferLoop()
+	go startWebService(webServicePort)
+	fmt.Printf("Web service started on %d port\n", webServicePort)
+	return &h
+}
+
+// Main loop for this handler, each log line will be sended to drain (if DrainURI specified) and copied to log storage
+func (h *Handler) ringBufferLoop() {
+	var err error
+	var projectName string
+	var message string
+	for {
+		m := h.Get()
+		if m == nil {
+			break
+		}
+		if h.DrainURI != "" {
+			drain.SendToDrain(m.String(), h.DrainURI)
+		}
+		message = m.String()
+		projectName, err = getProjectName(message)
+		if err != nil {
+			projectName = "unknown"
+			message = fmt.Sprintln(err)
+		}
+		addToStorage(projectName, message)
+	}
+	h.End()
+}
+
+// Only one http handler which process requests
+func httpHandler(w http.ResponseWriter, r *http.Request) {
+	if r.Method == "POST" {
+		match := regexpForPost.FindStringSubmatch(r.RequestURI)
+
+		if match == nil {
+			fmt.Fprintf(w, "Could not get application name from url: %s", r.RequestURI)
```

Comment:
I will look into that a little more.  An infinite loop when the logger writes a log message isn't what _should_ be happening.  (I'm not saying it can't happen-- in fact, I think I've seen it.  Only saying it _shouldn't_.)  When the logger writes log messages, they should be going to `STDOUT` and as in the case of all other containers, the `deis-logspout` agent running on the host `deis-logger` runs on should see that message and forward it _to_ `deis-logger`.  `deis-logger` will process that messages (that originated from itself), but it doesn't output (re-output in this case) log messages that is receives... the message should be treated just as is a message from any other component or app.

All that being said, there may be an opportunity to short-circuit that process because it's really inefficient.